### PR TITLE
[dagster-aws] make PipesCloudWatchMessageReader subclass PipesThreadedMessageReader

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

PipesCloudWatchMessageReader can now read messages & logs in background and read from streams which are not known initially.

This allows:
 - non-blocking `PipesCloudWatchMessageReader` usage
 - dynamically adding multiple of the new `PipesCloudWatchLogReader`s once the target log streams become known  

Existing Pipes client have been modified to follow this workflow.

## How I Tested These Changes
Added tests for new `PipesCloudWatchMessageReader` capabilties

## Changelog

- [x] [dagster-aws] `PipesCloudWatchMessageReader` now non-blocking (reads logs in background threads)

### CODE REVIEW

**Comment by danielgafni** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24809?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24809** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24809?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24469** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24469?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danielgafni and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py:187]: nit: consolidate access to `task["containers"]` up top and only have one pyright warning

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py
Comment (schrockn) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py:254]: yes please consolidate all these accesses

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/message_readers.py
Comment (danielgafni) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/message_readers.py:255]: @smackesey seems like if we do something like this the `can_start` method doesn't even need to have `params` argument.

We can set the actual attributes in `on_opened` and `on_launched`. 

No need to introduce another source of parameters. 

What do you think?

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py
Comment (smackesey) [inline python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py:45]: Not sure why annotation is removed here but you should explicitly put `Any` if you are changing it to `Any`

**Review by smackesey — APPROVED** [review]
So long as the new implementation here passes tests it should be good to merge (once the PRs it is stacked on are approved).

However, the summary is very thin. Can you please elaborate on the delta between the new and old functionality that this PR is adding? (e.g. How are these new streams that can be read detected?)

File: python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py:45]: Because `boto3.client` is an incorrect type annotation. 

`boto3.client("<service>") returns a dynamically generated class, it doesn't return anything like `boto.client`. 

Vanilla `boto3` is untyped (unless `boto3-stubs` is installed) so `Any` is indeed the best option we have here. 

Alternatively, we can do: 

```python
if TYPE_CHECKING:
    # requries boto3-stubs to be installed
    from mypy_boto3_stubs_ecs.client import ECSClient
    
...
    client: Optional["ECSCLient"] = None
```

**Pull request merged**
